### PR TITLE
VSCode: Add snapshot tests for Cairo grammar

### PR DESCRIPTION
**Pull request by mkaput**
This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5846)

### CODE REVIEW

**Review by orizi — CHANGES_REQUESTED** [review]
> Reviewed 3 of 98 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5846)* status: 3 of 98 files reviewed, 1 unresolved discussion (waiting on @mkaput)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/5846#-O-qBUZsA0M7ew5vGaTO:-O-qBUZsA0M7ew5vGaTP:bk5n6m2) (no related file):*
what is all that?

___

**Review by mkaput — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5846)* status: 3 of 98 files reviewed, 1 unresolved discussion (waiting on @orizi)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/5846#-O-qBUZsA0M7ew5vGaTO:-O-qCo8j2DhVumAxsLry:b-sddxxq) (no related file):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

what is all that?
</blockquote></details>

I took all code snippets from `cairo-lang-parser` tests and run vscode highlighter over them and saved results as fixtures.

The goal is to catch syntax changes and guard from forgetting to sync vscode grammar

___

**Review by mkaput — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5846)* status: 3 of 98 files reviewed, 1 unresolved discussion (waiting on @orizi)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/5846#-O-qBUZsA0M7ew5vGaTO:-O-qDXnG5Rq3dw63MWl_:b-4yybzo) (no related file):*
<details><summary><i>Previously, mkaput (Marek Kaput) wrote…</i></summary><blockquote>

I took all code snippets from `cairo-lang-parser` tests and run vscode highlighter over them and saved results as fixtures.

The goal is to catch syntax changes and guard from forgetting to sync vscode grammar
</blockquote></details>

The snapshot files are pretty verbose, hence so many lines added

___

**Review by orizi — COMMENTED** [review]
> Reviewed 34 of 98 files at r1.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5846)* status: 37 of 98 files reviewed, all discussions resolved

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/5846#-O-qBUZsA0M7ew5vGaTO:-O-qJjiB72R2naNx1edn:b27y052) (no related file):*
<details><summary><i>Previously, mkaput (Marek Kaput) wrote…</i></summary><blockquote>

The snapshot files are pretty verbose, hence so many lines added
</blockquote></details>

so now parser changes would need to run js testing?

___

**Review by mkaput — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5846)* status: 37 of 98 files reviewed, all discussions resolved (waiting on @orizi)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/5846#-O-qBUZsA0M7ew5vGaTO:-O09SdFWDmyVDLcK_m05:bt86qxy) (no related file):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

so now parser changes would need to run js testing?
</blockquote></details>

not parser itself, but its tests

the following command should do the job:
```
CAIRO_FIX_TESTS=1 npm test
```

___

**Review by orizi — APPROVED** [review]
> Reviewed 2 of 98 files at r1, 92 of 92 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5846)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @mkaput)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/5846#-O-qBUZsA0M7ew5vGaTO:-O09Z9PsEAa-l-cGuEJi:b-i1z6gj) (no related file):*
<details><summary><i>Previously, mkaput (Marek Kaput) wrote…</i></summary><blockquote>

not parser itself, but its tests

the following command should do the job:
```
CAIRO_FIX_TESTS=1 npm test
```
</blockquote></details>

that is probably even worse - these are test that we are going to completely ignore - but gonna hinder our ability to do anything new in the area. 

___

**Review by mkaput — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5846)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @mkaput)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/5846#-O-qBUZsA0M7ew5vGaTO:-O09_5xY2wfDPyvTClBk:bhvb8c3) (no related file):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

that is probably even worse - these are test that we are going to completely ignore - but gonna hinder our ability to do anything new in the area.
</blockquote></details>

Are you able to predict what of the following would be the pain points in this flow?

1. requiring Node.js to be installed on dev machine in case CI yells at the PR and this command would have to be executed
2. the fact that this command has to be run separately
3. some kind of lack of documentation?

I'd like to identify what should I optimize or if simply I should abandon this approach.

Abandoning would be OK for me provided that somebody (I?) will be notified on any language grammar change, to keep vscode grammar up-to-date. The goal of this PR is to avoid having maintainer attention which is risky IMO :/

___

**Review by orizi — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5846)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @mkaput)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/5846#-O-qBUZsA0M7ew5vGaTO:-O09gux6B4uUbp3iQvS8:bctme1h) (no related file):*
<details><summary><i>Previously, mkaput (Marek Kaput) wrote…</i></summary><blockquote>

Are you able to predict what of the following would be the pain points in this flow?

1. requiring Node.js to be installed on dev machine in case CI yells at the PR and this command would have to be executed
2. the fact that this command has to be run separately
3. some kind of lack of documentation?

I'd like to identify what should I optimize or if simply I should abandon this approach.

Abandoning would be OK for me provided that somebody (I?) will be notified on any language grammar change, to keep vscode grammar up-to-date. The goal of this PR is to avoid having maintainer attention which is risky IMO :/
</blockquote></details>

i like the concept in general - but my main fear is developer looking it these files and not knowing what happened at all.

if we just add this to our nightly ci (so we can respond slowly to it), it is fine with me.

___

**Review by mkaput — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5846)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @mkaput)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/5846#-O-qBUZsA0M7ew5vGaTO:-O0A0nvXEbqVoXOelxKA:bb6un1j) (no related file):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

i like the concept in general - but my main fear is developer looking it these files and not knowing what happened at all.

if we just add this to our nightly ci (so we can respond slowly to it), it is fine with me.
</blockquote></details>

Done

___

**Review by orizi — CHANGES_REQUESTED** [review]
:lgtm:

> Reviewed 12 of 12 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5846#-:-O0AW1_DCWQLYCMhPFXs:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @mkaput)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/5846#-O0AW1qrCNrf_9ynribn:-O0AW1qrCNrf_9ynribo:bm3uwni) (no related file):*
Please add someone for the .mjs stuff.

___

**Review by orizi — COMMENTED** [review]
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5846)* status: all files reviewed, 1 unresolved discussion (waiting on @mkaput)

File: vscode-cairo/test-support/grammar.mjs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      const tokenText = escapeWhitespace(line.slice(token.startIndex, token.endIndex)).padEnd(20);
+      const scopes = token.scopes.join(" ");
+      output.push(`${tokenText} ${scopes}`);
+    }
+  }
+  return output.join("\n");
+}
+
+function escapeWhitespace(str) {
+  return str.replace(/\n/g, "↩").replace(/\t/g, "⇥").replace(/\s/g, "␣");
```
Comment (Arcticae) [inline vscode-cairo/test-support/grammar.mjs:74]: couldn't this be done with one `replace` and more elaborate regex?

Comment (mkaput) [reply vscode-cairo/test-support/grammar.mjs]: Done

File: vscode-cairo/test/cairo.tmLanguage.mjs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import assert from "assert/strict";
+import test from "node:test";
+import corpus from "../test-support/cairo-corpus.gen.mjs";
```
[the remark this answers is not in the document]
Comment (mkaput) [reply vscode-cairo/test/cairo.tmLanguage.mjs]: Done

**Review by orizi — COMMENTED** [review]
> Reviewed 7 of 7 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5846)* status: all files reviewed, 3 unresolved discussions (waiting on @Arcticae and @mkaput)

**Review by Arcticae — COMMENTED** [review]
> Reviewed 2 of 98 files at r1, 89 of 92 files at r2, 1 of 1 files at r3, 5 of 12 files at r4, 7 of 7 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5846)* status: all files reviewed, 1 unresolved discussion (waiting on @mkaput)

**Comment by Arcticae** [comment]
:lgtm:

**Review by Arcticae — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5846)* status: all files reviewed, 1 unresolved discussion (waiting on @mkaput)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/5846#-O0AW1qrCNrf_9ynribn:-O0E0e7V5ZOIitjk7IS4:bg7igkv) (no related file):*
<details><summary><i>Previously, orizi wrote…</i></summary><blockquote>

Please add someone for the .mjs stuff.
</blockquote></details>

:lgtm: 

___

**Review by orizi — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/5846)* status: all files reviewed, 1 unresolved discussion (waiting on @mkaput)

**Pull request merged**